### PR TITLE
Add support to retry function execution on invocation failures

### LINKED ISSUE #6664: Add support for funciton execution retries
Consume WebJobs SDK changes made for https://github.com/Azure/azure-webjobs-sdk/issues/1597
 
Draft PR: https://github.com/Azure/azure-functions-host/pull/6603

### PR DESCRIPTION
**Pull request by pragnagopa**
### Issue describing the changes in this PR

Functions host changes to add retry support : https://github.com/Azure/azure-webjobs-sdk/pull/2463

Retry can be added in `host.json` for function app level or can be added in `function.json` for function level. If both are specified, `function.json` takes precedence.

```json
"retry": {
    "strategy": "fixedDelay",
    "maxRetryCount": 2,
    "delayInterval": "00:00:10"
  }
``` 
resolves #6664

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [x] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [ ] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #6684
* [x] I have added all required tests (Unit tests, E2E tests)

Note: Documentation for this feature is tracked internally.

### CODE REVIEW

File: src/WebJobs.Script/Description/FunctionDescriptorProvider.cs
Comment (mathewc) [inline src/WebJobs.Script/Description/FunctionDescriptorProvider.cs]: I think the host level option should be applied in ScriptHost.CreateTypeAttributes, so the attribute is applied at the class level

File: src/WebJobs.Script/WebJobs.Script.csproj
Comment (mathewc) [inline src/WebJobs.Script/WebJobs.Script.csproj]: This is just temp right?

File: src/WebJobs.Script.Abstractions/Description/Retry.cs
Comment (mathewc) [inline src/WebJobs.Script.Abstractions/Description/Retry.cs]: Wondering if we should have this strategy property. We could instead infer based on properties user has set, so they don't have to specify this? What do you think @brettsam ?

Comment (brettsam) [reply src/WebJobs.Script.Abstractions/Description/Retry.cs]: Then the rules would be:
1. If you explicitly set DelayInterval, it's Fixed
2. If you explicitly set Minimum or Maximum, then it's Exponential.

What if you don't set any? Or what if there's a conflict in the rules?

Comment (mathewc) [reply src/WebJobs.Script.Abstractions/Description/Retry.cs]: Yes, I was thinking something like that. Note that we still have similar validation work to do even we had them specify a strategy value, e.g. they might declare "fixedDelay" but not specify any other values. @pragnagopa how are you handling validation - where will it fail in cases like this?

Anyhow, maybe it's OK to just be explicit here and have a strategy property. Can we name it "retry" rather than "functionRetry" in function.json? If we ever supported functionTimeout specification in function.json, we'd also probably want to call that just "timeout", rather than "functionTimeout" as in host.json.

```javascript
"retry": {
  "strategy": "fixedDelay",
  "maxRetryCount": 3,
  "delayInterval": "00:00:10"
}

"retry": {
  "strategy": "exponentialBackoff",
  "maxRetryCount": 3,
  "minimumInterval": "00:00:02",
  "maximumInterval": "00:00:30"
}
```

File: sample/Node-Retry/HttpTrigger-RetryFunctionJson/function.json
Comment (mathewc) [inline sample/Node-Retry/HttpTrigger-RetryFunctionJson/function.json]: Formatting and casing off

File: sample/Node-Retry/host.json
Comment (mathewc) [inline sample/Node-Retry/host.json]: Formatting and casing off

File: src/WebJobs.Script/CustomAttributeBuilderUtility.cs
Comment (mathewc) [inline src/WebJobs.Script/CustomAttributeBuilderUtility.cs]: formatting off here and below

File: src/WebJobs.Script/Host/ScriptHost.cs
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: nit: spacing

File: test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs
Comment (mathewc) [inline test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs]: We're not verifying the host.json retry parsing anywhere are we?

File: src/WebJobs.Script.Abstractions/Description/Retry.cs
Comment (pragnagopa) [reply src/WebJobs.Script.Abstractions/Description/Retry.cs]: I prefer having `strategy` specified explicitly - bit more clear on which values to specify.  Validation for these values comes from underlying strategies - https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Host/Timers/RandomizedExponentialBackoffStrategy.cs and https://github.com/Azure/azure-webjobs-sdk/blob/dev/src/Microsoft.Azure.WebJobs.Host/Timers/FixedDelayStrategy.cs

File: test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs
Comment (pragnagopa) [reply test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs]: nope. I think validation built into webjobs sdk is enough.

File: src/WebJobs.Script/WebJobs.Script.csproj
Comment (pragnagopa) [reply src/WebJobs.Script/WebJobs.Script.csproj]: `FunctionMetadata` is now part of the `WebJobs.Script.Abstractions` . Udpated FunctionMetadata to include new property `Retry`

File: sample/Node-Retry/HttpTrigger-RetryFunctionJson/function.json
Comment (pragnagopa) [reply sample/Node-Retry/HttpTrigger-RetryFunctionJson/function.json]: fixed!

File: test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs
Comment (pragnagopa) [reply test/WebJobs.Script.Tests/FunctionMetadataProviderTests.cs]: Added validation - creating attributes earlier earlier so it fails sooner

**Comment by pragnagopa** [comment]
@mathewc / @brettsam  - PR ready for review!

@jeffhollan  - let me know if there is a doc I can link here.

File: src/WebJobs.Script.Abstractions/Description/Retry.cs
Comment (pragnagopa) [reply src/WebJobs.Script.Abstractions/Description/Retry.cs]: If `interval` is not specified, it defaults to `00:00:00` - is this ok? or should we enforce specifying non-zero delay for each strategy?

Comment (mathewc) [reply src/WebJobs.Script.Abstractions/Description/Retry.cs]: Let's require them to specify it explicitly. the underlying SDK FixedRetryAttribute requires a value to be explicitly provided on construction right? This would be consistent with that.

File: sample/NodeRetry/HttpTrigger-RetryFunctionJson/function.json
Comment (mathewc) [inline sample/NodeRetry/HttpTrigger-RetryFunctionJson/function.json]: fix casing (strategy, maxRetryCount, etc.)

File: sample/NodeRetry/host.json
Comment (mathewc) [inline sample/NodeRetry/host.json]: formatting casing off still :)

File: src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs
Comment (mathewc) [inline src/WebJobs.Script.Abstractions/Description/FunctionMetadata.cs]: nit: "Gets or sets the optional function execution retry strategy to use on invocation failures.<see cref="Abstractions.Description.Retry"/>"

File: src/WebJobs.Script/Config/ScriptJobHostOptions.cs
Comment (mathewc) [inline src/WebJobs.Script/Config/ScriptJobHostOptions.cs]: nit: "Gets or sets the retry options to use for failed function invocations."

File: src/WebJobs.Script.Abstractions/Description/Retry.cs
Comment (mathewc) [inline src/WebJobs.Script.Abstractions/Description/Retry.cs]: Thinking we should name this RetryOptions

File: src/WebJobs.Script/WebJobs.Script.csproj
Comment (mathewc) [reply src/WebJobs.Script/WebJobs.Script.csproj]: I mean the adding of the project reference is temporary, to allow you to easily test out your changes? You'll revert this before checking in, because we want a nuget package dependency, not a project dependency.

File: src/WebJobs.Script.Abstractions/Description/Retry.cs
Comment (pragnagopa) [reply src/WebJobs.Script.Abstractions/Description/Retry.cs]: Fixed!

File: src/WebJobs.Script/WebJobs.Script.csproj
Comment (pragnagopa) [reply src/WebJobs.Script/WebJobs.Script.csproj]: Ahh. Thanks. Makes sense now :) Yes will fix the PR once webjobs SDK is on myget

**Review by pragnagopa — COMMENTED** [review]
Addressed CR feedback!

File: src/WebJobs.Script/WebJobs.Script.csproj
Comment (pragnagopa) [reply src/WebJobs.Script/WebJobs.Script.csproj]: fixed references!

**Review by mathewc — APPROVED** [review]

**Comment by amotl** [comment]
Dear @pragnagopa and @mathewc,

we are currently looking into running full integration tests with [pytest](https://pytest.org/) through a host instance provided by [azure-functions-core-tools](https://github.com/Azure/azure-functions-core-tools). We are probably aiming at similar things which have been outlined within [1] and [2] and already sketched out within [3] and [4] by @jeffhollan and @Francisco-Gamino, but this time for a pure Python environment.

As we don't only want to run tests looking for successful outcomes but also want to wrap the test harness around erroneous function invocations, we have been looking for a way to get around the default setting with a retry count of 3, which terminates the host instance after trying to invoke the function three times like
```
Exceeded language worker restart retry count for runtime:python. Shutting down and proactively recycling the Functions Host to recover
```

Actually, we would like to reduce that to "1" when running within the test harness in order to speed things up.

So, we have been happy to find things like [host.json](https://github.com/Azure/azure-functions-host/blob/dd062a5cc9e1f8e8bb875d3e6068b1df3865257a/sample/NodeRetry/host.json) or [function.json](https://github.com/Azure/azure-functions-host/blob/dd062a5cc9e1f8e8bb875d3e6068b1df3865257a/sample/NodeRetry/HttpTrigger-RetryFunctionJson/function.json) just recently coming from your pen.

```
  "retry": {
    "strategy": "fixedDelay",
    "maxRetryCount": 2,
    "delayInterval": "00:00:03"
  }
```

Now - as these new attributes are not mentioned within [host.json reference for Azure Functions 2.x and later](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json) yet - am I right to assume the corresponding improvements are not part of the most recent release of _Azure Functions Core Tools_ yet? I am currently running
```
$ func --version
3.0.2931
```

Regarding this, I have two questions:
- Would this be resolved with #6684? 
- Will I have to wait until the next official release of _Azure Functions Core Tools_?
- Is there a way to compile this machinery on behalf of my own or is there way to download nightly builds of https://github.com/Azure/azure-functions-core-tools/releases?

Please correct me if I am wrong on any of the details I was trying to outline here.

With kind regards,
Andreas.

[1] https://stackoverflow.com/questions/62749060/unit-and-integration-test-for-azure-function-with-servicebustrigger
[2] https://medium.com/cheranga/testing-azure-functions-part-2-86e036785f59
[3] https://github.com/jeffhollan/functions-test-helper
[4] https://github.com/Azure/azure-functions-integration-tests

**Comment by jeffhollan** [comment]
@pragnagopa to help - but I'm not sure if the retry policy here is going to apply to the retry you are showing there (which is retrying communication to a language worker, and not just retrying execution of a message).  That said, to answer other questions, yes it's not yet in core tools, docs will merge sometime next week once feature is live and core tools get released.  Likely some way to test out with latest WebJobs SDK and such but I'm not sure how. The release is rolling out right now I believe.

**Comment by pragnagopa** [comment]
@amotl  - @jeffhollan  is right.  Retry support added in this PR is for function executions. Following log message

```
Exceeded language worker restart retry count for runtime:python. Shutting down and proactively recycling the Functions Host to recover
```

is coming from built-in retries to handle process crashes - in this case if python process crashes - functions host retries upto `3` times. We do not intend to expose this as a configuration setting for end-users.

**Comment by amotl** [comment]
Dear @jeffhollan and @pragnagopa,

thanks a bunch for your quick answers. What I was trying to achieve is to make the complete function host instance terminate itself by doing such a thing:

```
def main(events: List[func.EventHubEvent]):

    # Process a bunch of events here.
    for event in events:
        pass

    # Exit the function host when running SuT.
    # https://docs.pytest.org/en/latest/example/simple.html#pytest-current-test-environment-variable
    if "PYTEST_CURRENT_TEST" in os.environ:
        sys.exit(0)
```

In this manner, our idea was to spin up a function/host instance from a pytest fixture, submit a single or more messages to it by e.g. invoking
```
http http://0.0.0.0:7071/admin/functions/eventHubTrigger input=foobar
```

and then check the outcome of that what has been done when processing the message(s) (e.g. writing to a database).

After seeing that the host was shutting down itself after three attempts, we figured there would be a way to get around terminating the process from outside (the pytest fixture) after the test run has completed.

Now, after learning what the added "retry" functionality is all about, I believe this was an ill-fated idea.

With kind regards,
Andreas.

P.S.: One last question, @jeffhollan: I see that you are terminating the function host after 30 seconds like [`Host.StartAsync().Wait(TimeSpan.FromSeconds(30));`](https://github.com/jeffhollan/functions-test-helper/blob/7caa7af6c9e7ad3ce5199a405db01dd3b4c24ad1/FunctionTestHelper/EndToEndTestFixture.cs#L60), right? This might be exactly the spot where I was aiming at: You just lazily wait there for 30 seconds and then shut down the whole machinery? 

In contrast to this, I am aiming at getting this together more tightly and terminate the function host **directly** after it has done its job processing all the messages coming from a **single function invocation**.

If you know any other people or resources having attempted similar things running full integration tests through _Core Tools' Function Host_ which could guide me along, please let me know. We are aiming to wrap a test harness around a scenario where messages are ingested from EventHub (probably similar to your example at [1]) and submitted to both CrateDB and Kafka, see also https://github.com/Azure/azure-functions-kafka-extension/issues/181.

[1] https://github.com/jeffhollan/functions-test-helper/blob/master/FunctionApp.Tests.Integration/EventEndToEndTests.cs

**Comment by pragnagopa** [comment]
@amotl - looking at the code you shared:

```
if "PYTEST_CURRENT_TEST" in os.environ:
        sys.exit(0)
```

`sys.exit(0)` is causing the python process to exit - which triggers functions host to restart python language worker process. 

Please open new issue on : https://github.com/Azure/azure-functions-python-worker/issues if you need more guidance how to proceed.

**Comment by amotl** [comment]
Dear Pragna,

thanks for your answer and sorry for spamming this PR. I believe we will just use [`terminate()`](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.terminate) in order to kill the function host process from outside within the test runner.

Regarding another question to proceed where we are aiming at, I have just created https://github.com/Azure/azure-functions-python-worker/issues/764.

Thanks again and with kind regards,
Andreas.

**Pull request merged**
